File: CS/CustomTodayExample/MyDateEdit.cs
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using System.ComponentModel;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Controls;
using System;
using DevExpress.XtraEditors.Calendar;
using System.Drawing;

namespace DXSample {
    public class MyDateEdit :DateEdit {
        static MyDateEdit() { RepositoryItemMyDateEdit.RegisterMyDateEdit(); }
        public MyDateEdit() : base() { }

        public override string EditorTypeName { get { return RepositoryItemMyDateEdit.MyDateEditName; } }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyDateEdit Properties { get { return (RepositoryItemMyDateEdit)base.Properties; } }

        protected override PopupBaseForm CreatePopupForm() {
            if (Properties.IsVistaDisplayModeInternal) return new MyVistaPopupDateEditForm(this);
            return new MyPopupDateEditForm(this);
        }
    }

    [UserRepositoryItem("RegisterMyDateEdit")]
    public class RepositoryItemMyDateEdit :RepositoryItemDateEdit {
        static RepositoryItemMyDateEdit() { RegisterMyDateEdit(); }
        public RepositoryItemMyDateEdit() : base() { }

        internal const string MyDateEditName = "MyDateEdit";
        public override string EditorTypeName { get { return MyDateEditName; } }

        public static void RegisterMyDateEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MyDateEditName, typeof(MyDateEdit), typeof(RepositoryItemMyDateEdit),
                typeof(DateEditViewInfo), new ButtonEditPainter(), true));
        }

        internal bool IsVistaDisplayModeInternal { get { return IsVistaDisplayMode(); } }

        static readonly object fCustomToday = new object();
        public event EventHandler<CustomTodayEventArgs> CustomToday {
            add { Events.AddHandler(fCustomToday, value); }
            remove { Events.RemoveHandler(fCustomToday, value); }
        }
        public DateTime GetCustomToday() {
            EventHandler<CustomTodayEventArgs> handler = (EventHandler<CustomTodayEventArgs>)Events[fCustomToday];
            if (handler == null) return DateTime.Today;
            CustomTodayEventArgs args = new CustomTodayEventArgs();
            handler(this, args);
            return args.Today;
        }

        public override void Assign(RepositoryItem item) {
            base.Assign(item);
            BeginUpdate();
            try {
                RepositoryItemMyDateEdit source = (RepositoryItemMyDateEdit)item;
                Events.AddHandler(fCustomToday, source.Events[fCustomToday]);
            } finally { EndUpdate(); }
        }
    }

    public class MyPopupDateEditForm :PopupDateEditForm {
        public MyPopupDateEditForm(MyDateEdit ownerEdit) : base(ownerEdit) { }

        public new MyDateEdit OwnerEdit { get { return (MyDateEdit)base.OwnerEdit; } }

        protected override DateEditCalendar CreateCalendar() {
            return new MyDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue);
        }

        protected override void ResetState() {
            DateTime date = OwnerEdit.DateTime;
            if (date == DateTime.MinValue) date = OwnerEdit.Properties.GetCustomToday();
            Calendar.ResetState(OwnerEdit.EditValue, date);
        }
    }

    public class MyVistaPopupDateEditForm :VistaPopupDateEditForm {
        public MyVistaPopupDateEditForm(MyDateEdit ownerEdit) : base(ownerEdit) { }

        public new MyDateEdit OwnerEdit { get { return (MyDateEdit)base.OwnerEdit; } }

        protected override DateEditCalendar CreateCalendar() {
            VistaDateEditCalendar result = new MyVistaDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue);
            result.OkClick += OnOkClick;
            return result;
        }

        protected override void Dispose(bool disposing) {
            if (disposing && Calendar != null)
                ((VistaDateEditCalendar)Calendar).OkClick -= OnOkClick;
            base.Dispose(disposing);
        }

        protected override void ResetState() {
            DateTime date = OwnerEdit.DateTime;
            if (date == DateTime.MinValue) date = OwnerEdit.Properties.GetCustomToday();
            Calendar.ResetState(OwnerEdit.EditValue, date);
        }
    }

    public class MyDateEditCalendar :DateEditCalendar {
        public MyDateEditCalendar(RepositoryItemMyDateEdit item, object editDate) : base(item, editDate) { }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyDateEdit Properties { get { return (RepositoryItemMyDateEdit)base.Properties; } }

        protected override void OnTodayButtonClick(object sender, EventArgs e) {
            DateTime today = Properties.GetCustomToday();
            SetSelection(today);
            OnDateTimeCommit(today, false);
        }

        protected override DateEditPainter CreatePainter() {
            return new MyDateEditPainter(this);
        }
    }

    public class MyVistaDateEditCalendar :VistaDateEditCalendar {
        public MyVistaDateEditCalendar(RepositoryItemMyDateEdit item, object editDate) : base(item, editDate) { }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyDateEdit Properties { get { return (RepositoryItemMyDateEdit)base.Properties; } }

        protected override void OnTodayButtonClick(object sender, EventArgs e) {
            DateTime today = Properties.GetCustomToday();
            SetSelection(today);
            OnDateTimeCommit(today, false);
        }

        protected override DateEditPainter CreatePainter() {
            return new MyVistaDateEditPainter(this);
        }

        protected override DateEditInfoArgs CreateInfoArgs() {
            return new MyVistaDateEditInfoArgs(this);
        }
    }

    public class CustomTodayEventArgs :EventArgs {
        private DateTime fToday = DateTime.Today;
        public DateTime Today {
            get { return fToday; }
            set { fToday = value; }
        }
    }

    public class MyDateEditPainter :DateEditPainter {
        public MyDateEditPainter(DateEditCalendarBase calendar) : base(calendar) { }

        protected override void DrawDayCell(CalendarObjectInfoArgs info, DayNumberCellInfo cell) {
            cell.Today = cell.Date == fToday;
            base.DrawDayCell(info, cell);
        }

        DateTime fToday;
        protected override void DrawMonthNumbers(CalendarObjectInfoArgs info) {
            fToday = ((MyDateEditCalendar)Calendar).Properties.GetCustomToday();
            base.DrawMonthNumbers(info);
        }
    }

    public class MyVistaDateEditPainter :VistaDateEditPainter {
        public MyVistaDateEditPainter(DateEditCalendarBase calendar) : base(calendar) { }

        protected override void DrawDayCell(CalendarObjectInfoArgs info, DayNumberCellInfo cell) {
            cell.Today = cell.Date == fToday;
            base.DrawDayCell(info, cell);
        }

        DateTime fToday;
        protected override void DrawMonthNumbers(CalendarObjectInfoArgs info) {
            fToday = ((MyVistaDateEditCalendar)Calendar).Properties.GetCustomToday();
            base.DrawMonthNumbers(info);
        }
    }

    public class MyVistaDateEditInfoArgs :VistaDateEditInfoArgs {
        public MyVistaDateEditInfoArgs(DateEditCalendarBase calendar) : base(calendar) { }

        DateTime fToday;
        protected override void CalcDayNumberCells() {
            fToday = ((MyVistaDateEditCalendar)Calendar).Properties.GetCustomToday();
            base.CalcDayNumberCells();
        }

        protected override void UpdateCellAppearance(DayNumberCellInfo cell) {
            cell.Today = fToday == cell.Date;
            base.UpdateCellAppearance(cell);
        }
    }
}

File: VB/CustomTodayExample/MyDateEdit.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Controls
Imports System
Imports DevExpress.XtraEditors.Calendar
Imports System.Drawing

Namespace DXSample
	Public Class MyDateEdit
		Inherits DateEdit
		Shared Sub New()
			RepositoryItemMyDateEdit.RegisterMyDateEdit()
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMyDateEdit.MyDateEditName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreatePopupForm() As PopupBaseForm
			If Properties.IsVistaDisplayModeInternal Then
				Return New MyVistaPopupDateEditForm(Me)
			End If
			Return New MyPopupDateEditForm(Me)
		End Function
	End Class

	<UserRepositoryItem("RegisterMyDateEdit")> _
	Public Class RepositoryItemMyDateEdit
		Inherits RepositoryItemDateEdit
		Shared Sub New()
			RegisterMyDateEdit()
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Friend Const MyDateEditName As String = "MyDateEdit"
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return MyDateEditName
			End Get
		End Property

		Public Shared Sub RegisterMyDateEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MyDateEditName, GetType(MyDateEdit), GetType(RepositoryItemMyDateEdit), GetType(DateEditViewInfo), New ButtonEditPainter(), True))
		End Sub

		Friend ReadOnly Property IsVistaDisplayModeInternal() As Boolean
			Get
				Return IsVistaDisplayMode()
			End Get
		End Property

		Private Shared ReadOnly fCustomToday As Object = New Object()
		Public Custom Event CustomToday As EventHandler(Of CustomTodayEventArgs)
			AddHandler(ByVal value As EventHandler(Of CustomTodayEventArgs))
				Events.AddHandler(fCustomToday, value)
			End AddHandler
			RemoveHandler(ByVal value As EventHandler(Of CustomTodayEventArgs))
				Events.RemoveHandler(fCustomToday, value)
			End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As CustomTodayEventArgs)
            End RaiseEvent
		End Event
		Public Function GetCustomToday() As DateTime
			Dim handler As EventHandler(Of CustomTodayEventArgs) = CType(Events(fCustomToday), EventHandler(Of CustomTodayEventArgs))
			If handler Is Nothing Then
				Return DateTime.Today
			End If
			Dim args As New CustomTodayEventArgs()
			handler(Me, args)
			Return args.Today
		End Function

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			MyBase.Assign(item)
			BeginUpdate()
			Try
				Dim source As RepositoryItemMyDateEdit = CType(item, RepositoryItemMyDateEdit)
				Events.AddHandler(fCustomToday, source.Events(fCustomToday))
			Finally
				EndUpdate()
			End Try
		End Sub
	End Class

	Public Class MyPopupDateEditForm
		Inherits PopupDateEditForm
		Public Sub New(ByVal ownerEdit As MyDateEdit)
			MyBase.New(ownerEdit)
		End Sub

		Public Shadows ReadOnly Property OwnerEdit() As MyDateEdit
			Get
				Return CType(MyBase.OwnerEdit, MyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreateCalendar() As DateEditCalendar
			Return New MyDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue)
		End Function

		Protected Overrides Sub ResetState()
			Dim [date] As DateTime = OwnerEdit.DateTime
			If [date] = DateTime.MinValue Then
				[date] = OwnerEdit.Properties.GetCustomToday()
			End If
			Calendar.ResetState(OwnerEdit.EditValue, [date])
		End Sub
	End Class

	Public Class MyVistaPopupDateEditForm
		Inherits VistaPopupDateEditForm
		Public Sub New(ByVal ownerEdit As MyDateEdit)
			MyBase.New(ownerEdit)
		End Sub

		Public Shadows ReadOnly Property OwnerEdit() As MyDateEdit
			Get
				Return CType(MyBase.OwnerEdit, MyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreateCalendar() As DateEditCalendar
			Dim result As VistaDateEditCalendar = New MyVistaDateEditCalendar(OwnerEdit.Properties, OwnerEdit.EditValue)
            AddHandler result.OkClick, AddressOf OnOkClick
			Return result
		End Function

		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing AndAlso Calendar IsNot Nothing Then
                RemoveHandler CType(Calendar, VistaDateEditCalendar).OkClick, AddressOf OnOkClick
			End If
			MyBase.Dispose(disposing)
		End Sub

		Protected Overrides Sub ResetState()
			Dim [date] As DateTime = OwnerEdit.DateTime
			If [date] = DateTime.MinValue Then
				[date] = OwnerEdit.Properties.GetCustomToday()
			End If
			Calendar.ResetState(OwnerEdit.EditValue, [date])
		End Sub
	End Class

	Public Class MyDateEditCalendar
		Inherits DateEditCalendar
		Public Sub New(ByVal item As RepositoryItemMyDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
		End Sub

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Sub OnTodayButtonClick(ByVal sender As Object, ByVal e As EventArgs)
			Dim today As DateTime = Properties.GetCustomToday()
			SetSelection(today)
			OnDateTimeCommit(today, False)
		End Sub

		Protected Overrides Function CreatePainter() As DateEditPainter
			Return New MyDateEditPainter(Me)
		End Function
	End Class

	Public Class MyVistaDateEditCalendar
		Inherits VistaDateEditCalendar
		Public Sub New(ByVal item As RepositoryItemMyDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
		End Sub

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Sub OnTodayButtonClick(ByVal sender As Object, ByVal e As EventArgs)
			Dim today As DateTime = Properties.GetCustomToday()
			SetSelection(today)
			OnDateTimeCommit(today, False)
		End Sub

		Protected Overrides Function CreatePainter() As DateEditPainter
			Return New MyVistaDateEditPainter(Me)
		End Function

		Protected Overrides Function CreateInfoArgs() As DateEditInfoArgs
			Return New MyVistaDateEditInfoArgs(Me)
		End Function
	End Class

	Public Class CustomTodayEventArgs
		Inherits EventArgs
		Private fToday As DateTime = DateTime.Today
		Public Property Today() As DateTime
			Get
				Return fToday
			End Get
			Set(ByVal value As DateTime)
				fToday = value
			End Set
		End Property
	End Class

	Public Class MyDateEditPainter
		Inherits DateEditPainter
		Public Sub New(ByVal calendar As DateEditCalendarBase)
			MyBase.New(calendar)
		End Sub

		Protected Overrides Sub DrawDayCell(ByVal info As CalendarObjectInfoArgs, ByVal cell As DayNumberCellInfo)
			cell.Today = cell.Date = fToday
			MyBase.DrawDayCell(info, cell)
		End Sub

		Private fToday As DateTime
		Protected Overrides Sub DrawMonthNumbers(ByVal info As CalendarObjectInfoArgs)
			fToday = (CType(Calendar, MyDateEditCalendar)).Properties.GetCustomToday()
			MyBase.DrawMonthNumbers(info)
		End Sub
	End Class

	Public Class MyVistaDateEditPainter
		Inherits VistaDateEditPainter
		Public Sub New(ByVal calendar As DateEditCalendarBase)
			MyBase.New(calendar)
		End Sub

		Protected Overrides Sub DrawDayCell(ByVal info As CalendarObjectInfoArgs, ByVal cell As DayNumberCellInfo)
			cell.Today = cell.Date = fToday
			MyBase.DrawDayCell(info, cell)
		End Sub

		Private fToday As DateTime
		Protected Overrides Sub DrawMonthNumbers(ByVal info As CalendarObjectInfoArgs)
			fToday = (CType(Calendar, MyVistaDateEditCalendar)).Properties.GetCustomToday()
			MyBase.DrawMonthNumbers(info)
		End Sub
	End Class

	Public Class MyVistaDateEditInfoArgs
		Inherits VistaDateEditInfoArgs
		Public Sub New(ByVal calendar As DateEditCalendarBase)
			MyBase.New(calendar)
		End Sub

		Private fToday As DateTime
		Protected Overrides Sub CalcDayNumberCells()
			fToday = (CType(Calendar, MyVistaDateEditCalendar)).Properties.GetCustomToday()
			MyBase.CalcDayNumberCells()
		End Sub

		Protected Overrides Sub UpdateCellAppearance(ByVal cell As DayNumberCellInfo)
			cell.Today = fToday = cell.Date
			MyBase.UpdateCellAppearance(cell)
		End Sub
	End Class
End Namespace